source/common/json/json_streamer.h

```diff
@@ -0,0 +1,261 @@
+#pragma once
+
+#include <memory>
+#include <stack>
+#include <string>
+
+#include "envoy/buffer/buffer.h"
+#include "envoy/common/optref.h"
+
+#include "absl/container/flat_hash_map.h"
+#include "absl/strings/string_view.h"
+#include "absl/types/variant.h"
+
+namespace Envoy {
+namespace Json {
+
+/**
+ * Provides an API for streaming JSON output, as an alternative to populating a
+ * JSON structure with an image of what you want to serialize, or using a
+ * protobuf with reflection. The advantage of this approach is that it does not
+ * require building an intermediate data structure with redundant copies of all
+ * strings, maps, and arrays.
+ */
+class Streamer {
```

Comment:
This whole structure seems pretty error-prone and dangerous, relying on the owner to keep ownership of the components and destroy/close them in the right order.

Did you consider implementing it as a state machine, such that it can keep track of its own levels internally, with an interface that you'd use like...
```
  streamer_.addMapStart();
  streamer_.addKey("foo");
  streamer_.addArrayStart();
  streamer_.addValue("bar");
  streamer_.endArray();
```
This could have an auto close everything on the destructor, and asserts for operations that don't apply in the current state, making the overall interface similar but without needing awkward member variables like json_histogram_map_2_, without needing to correctly order a bundle of destructors, and without any need for the "is_closed_" internal tracker.

Reply:
Per chat, this is somewhat a matter of taste but I think it's easier to read the call-sites if the maps and arrays are explicit. It also mirrors better what the programming model was when we were createing an entire shadow protobuf structure and then serializing it.